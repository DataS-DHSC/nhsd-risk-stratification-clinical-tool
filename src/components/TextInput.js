import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Fieldset, Input } from 'nhsuk-react-components';

const TextInput = ({ field, required, prevData, groupName, setField }) => {
  const { label, hint, fieldValidator } = field;

  const [localValue, setLocalValue] = useState(prevData[groupName]);

  const handleOnInput = (e) => {
    if (fieldValidator) {
      const errorMessage = fieldValidator(e.target.value);
      e.target.setCustomValidity(errorMessage);
    }
  };

  const onChangeHandler = (e) => {
    setLocalValue(e.target.value);
    setField(groupName, e.target.value);
  };

  const onBlurHandler = (e) => {
    if (e.target.value) {
      setLocalValue(e.target.value.trim());
      setField(groupName, e.target.value.trim());
    }
  };

  return (
    <Fieldset>
      <Fieldset.Legend className="hidden">{label}</Fieldset.Legend>

      <label
        htmlFor={groupName}
        className="nhsuk-heading-xs nhsuk-u-font-weight-bold nhsuk-u-margin-bottom-2"
        id={`${field.label.toLowerCase().replace(/ /g, '-')}-radios-label`}
      >
        {label}
      </label>

      <Input
        className="nhsuk-input nhsuk-input--width-10"
        autoComplete="off"
        id={groupName}
        name={groupName}
        type="text"
        hint={hint}
        value={localValue}
        onChange={onChangeHandler}
        onBlur={onBlurHandler}
        required={required}
        onInput={handleOnInput}
      />
    </Fieldset>
  );
};

TextInput.propTypes = {
  field: PropTypes.shape().isRequired,
  prevData: PropTypes.shape(),
  groupName: PropTypes.string.isRequired,
  required: PropTypes.bool,
  setField: PropTypes.func.isRequired,
};

TextInput.defaultProps = {
  prevData: {},
  required: false,
};

export default TextInput;
