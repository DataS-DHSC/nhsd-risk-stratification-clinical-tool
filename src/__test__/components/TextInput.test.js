import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import TextInput from '../../components/TextInput';

const defaultProps = {
  required: true,
  prevData: { testField: 'testValue' },
  groupName: 'testField',
  setField: jest.fn(),
};

describe('TextInput', () => {
  it('should match snapshot', () => {
    const props = {
      ...defaultProps,
      field: {
        label: 'test label',
      },
    };

    const container = render(<TextInput {...props} />);
    expect(container.asFragment()).toMatchSnapshot();
  });

  it('sets customValidity as soon as it recieves an onInput event', () => {
    const mockValidator = jest.fn();
    mockValidator.mockImplementationOnce(() => 'testValidationMessage');

    const props = {
      ...defaultProps,
      field: {
        label: 'test label',
        hint: 'test hint',
        fieldValidator: mockValidator,
      },
    };

    const { getByRole } = render(<TextInput {...props} />);
    const setCustomValidity = jest.fn();

    fireEvent.input(getByRole('textbox'), {
      target: { value: 'changed value', setCustomValidity },
    });

    expect(mockValidator).toHaveBeenCalledWith('changed value');
    expect(setCustomValidity).toHaveBeenCalledWith('testValidationMessage');
  });

  it('sets context data on input change', () => {
    const setField = jest.fn();

    const props = {
      ...defaultProps,
      field: {
        label: 'test label',
      },
      setField,
    };

    const { getByRole, asFragment } = render(<TextInput {...props} />);

    fireEvent.change(getByRole('textbox'), { target: { value: 'test value' } });

    expect(setField).toHaveBeenCalledWith('testField', 'test value');
    expect(asFragment()).toMatchSnapshot();
  });

  test.each([
    ['   test', 'test'],
    ['test   ', 'test'],
    ['  test  ', 'test'],
    ['test', 'test'],
    ['   ', ''],
  ])('should trim value onBlur', (input, trimmed) => {
    const setField = jest.fn();

    const props = {
      ...defaultProps,
      field: {
        label: 'test label',
      },
      setField,
    };

    const { getByRole } = render(<TextInput {...props} />);

    fireEvent.blur(getByRole('textbox'), { target: { value: input } });

    expect(setField).toHaveBeenCalledWith('testField', trimmed);
  });
});
