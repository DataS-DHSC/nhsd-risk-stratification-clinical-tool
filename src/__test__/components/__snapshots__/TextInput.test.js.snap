// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput sets context data on input change 1`] = `
<DocumentFragment>
  <div
    class="nhsuk-form-group"
  >
    <fieldset
      class="nhsuk-fieldset"
    >
      <legend
        class="nhsuk-fieldset__legend hidden"
      >
        test label
      </legend>
      <label
        class="nhsuk-heading-xs nhsuk-u-font-weight-bold nhsuk-u-margin-bottom-2"
        for="testField"
        id="test-label-radios-label"
      >
        test label
      </label>
      <div
        class="nhsuk-form-group"
      >
        <input
          autocomplete="off"
          class="nhsuk-input nhsuk-input nhsuk-input--width-10"
          id="testField"
          name="testField"
          required=""
          type="text"
          value="test value"
        />
      </div>
    </fieldset>
  </div>
</DocumentFragment>
`;

exports[`TextInput should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="nhsuk-form-group"
  >
    <fieldset
      class="nhsuk-fieldset"
    >
      <legend
        class="nhsuk-fieldset__legend hidden"
      >
        test label
      </legend>
      <label
        class="nhsuk-heading-xs nhsuk-u-font-weight-bold nhsuk-u-margin-bottom-2"
        for="testField"
        id="test-label-radios-label"
      >
        test label
      </label>
      <div
        class="nhsuk-form-group"
      >
        <input
          autocomplete="off"
          class="nhsuk-input nhsuk-input nhsuk-input--width-10"
          id="testField"
          name="testField"
          required=""
          type="text"
          value="testValue"
        />
      </div>
    </fieldset>
  </div>
</DocumentFragment>
`;
